### ISSUE TITLE
Repeated ServerWriter<C>->Write() gives API misuse of type GRPC_CALL_ERROR_TOO_MANY_OPERATIONS 

### ISSUE THREAD

**Issue by RubenGarcia** [grpc/grpc#38106]
<!--
### What version of gRPC and what language are you using?
v1.64.2, c++

### What operating system (Linux, Windows,...) and version?
Android TIRAMISU

### What runtime / compiler are you using (e.g. python version or version of gcc)
ndk\26.1.10909125 
clang.exe 
 clang version 17.0.2 

### What did you do?
Please provide either 1) A unit test for reproducing the bug or 2) Specific steps for us to follow to reproduce the bug. If there’s not enough information to debug the problem, gRPC team may close the issue at their discretion. You’re welcome to re-open the issue once you have a reproduction.

Create a server with an rpc which returns a stream
Implement the rpc following
https://grpc.io/docs/languages/cpp/basics/
like this

```
Status A::B(ServerWriter< C>* (ServerContext* context, const SubscriptionConfig* request,
                                                          ServerWriter< C>* writer) {
    threads.push_back(std::thread([this, writer] {
            respond(writer);
}));
  
    return Status::OK;

void A::respond(ServerWriter< C>* writer) {
  grpc::WriteOptions	options;
   while(true){
        A update;
        bool res = writer->Write(update, options);
        if (!res) {
            return;
        }
        std::this_thread::sleep_for(std::chrono::milliseconds(33));
 }
### What did you expect to see?

The server sends updates until the client closes the connection. 
https://grpc.github.io/grpc/cpp/classgrpc_1_1_server_writer.html
indicates
```
Synchronous (blocking) server-side API for doing for doing a server-streaming RPCs, where the outgoing message stream coming from the server has messages of type W
```
So I understand the function returns once the processing is done and there is nothing I need to do before sending another message.
The documentation also does not indicate any preconditions to call the function, nor it indicates any API usage instructions.

### What did you see instead?

Crash, with this in the log
grpc api misuse of type GRPC_CALL_ERROR_TOO_MANY_OPERATIONS

```
11-12 14:27:14.087  5915  6906 E GRPC    : call_op_set.h:977] API misuse of type GRPC_CALL_ERROR_TOO_MANY_OPERATIONS observed
11-12 14:27:14.088  5915  6906 F native  : F0000 00:00:1731418034.087504    6906 call_op_set.h:979] Check failed: false
11-12 14:27:14.088  5915  6906 F native  : terminating.
11-12 14:27:14.088  5915  6906 F native  : F0000 00:00:1731418034.087504    6906 call_op_set.h:979] Check failed: false
11-12 14:27:14.088  5915  6906 F native  : terminating.
--------- beginning of crash
11-12 14:27:14.089  5915  6906 F libc    : Fatal signal 6 (SIGABRT), code -1 (SI_QUEUE) in tid 6906 (com.appname), pid 5915 (appname)
11-12 14:27:14.190  6909  6909 I crash_dump64: obtaining output fd from tombstoned, type: kDebuggerdTombstoneProto
11-12 14:27:14.191  1180  1180 I tombstoned: received crash request for pid 6906
11-12 14:27:14.192  6909  6909 I crash_dump64: performing dump of process 5915 (target tid = 6906)
11-12 14:27:14.280  2660  2761 I GestureDetector: handleMessage TAP
11-12 14:27:14.281  2660  2660 I GestureDetector: handleMessage TAP
11-12 14:27:14.298  4206  5081 D BtGatt.GattService: onScanResult to scannerId: 5- eventType=0x1a, addressType=1, address=XX:XX:XX:XX:A8:42, primaryPhy=1, secondaryPhy=0, advertisingSid=0xff, txPower=127, rssi=-73, periodicAdvInt=0x0
11-12 14:27:14.309  4206  5081 D BtGatt.GattService: onScanResult to scannerId: 5- eventType=0x1a, addressType=1, address=XX:XX:XX:XX:D4:B7, primaryPhy=1, secondaryPhy=0, advertisingSid=0xff, txPower=127, rssi=-68, periodicAdvInt=0x0
11-12 14:27:14.488  1095  1095 D io_stats: !@   8,0 r 1190165 62959150 w 3528521 63905076 d 1002601 66585752 f 0 500473 iot 3283968 0 th 0 0 0 pt 0 inp 0 0 391579.624

11-12 14:27:14.974  5137  5267 I PolicyManager: [#CMH#] Is foreground game app: false
11-12 14:27:15.044  6909  6909 F DEBUG   : *** *** *** *** *** *** *** *** *** *** *** *** *** *** *** ***
11-12 14:27:15.044  6909  6909 F DEBUG   : Build fingerprint: 'samsung/b0quew/b0q:14/UP1A.231005.007/S908U1UES6EXG8:user/release-keys'
11-12 14:27:15.044  6909  6909 F DEBUG   : Revision: '13'
11-12 14:27:15.044  6909  6909 F DEBUG   : ABI: 'arm64'
11-12 14:27:15.044  6909  6909 F DEBUG   : Processor: '1'
11-12 14:27:15.044  6909  6909 F DEBUG   : Timestamp: 2024-11-12 14:27:14.211096138+0100
11-12 14:27:15.045  6909  6909 F DEBUG   : Process uptime: 206s
11-12 14:27:15.045  6909  6909 F DEBUG   : Cmdline: com.appname
11-12 14:27:15.045  6909  6909 F DEBUG   : pid: 5915, tid: 6906, name: appname  >>> com.appname <<<
11-12 14:27:15.045  6909  6909 F DEBUG   : uid: 10384
11-12 14:27:15.045  6909  6909 F DEBUG   : tagged_addr_ctrl: 0000000000000001 (PR_TAGGED_ADDR_ENABLE)
11-12 14:27:15.045  6909  6909 F DEBUG   : pac_enabled_keys: 000000000000000f (PR_PAC_APIAKEY, PR_PAC_APIBKEY, PR_PAC_APDAKEY, PR_PAC_APDBKEY)
11-12 14:27:15.045  6909  6909 F DEBUG   : signal 6 (SIGABRT), code -1 (SI_QUEUE), fault addr --------
11-12 14:27:15.045  6909  6909 F DEBUG   : Abort message: 'F0000 00:00:1731418034.087504    6906 call_op_set.h:979] Check failed: false '
11-12 14:27:15.045  6909  6909 F DEBUG   :     x0  0000000000000000  x1  0000000000001afa  x2  0000000000000006  x3  0000007bba965040
11-12 14:27:15.045  6909  6909 F DEBUG   :     x4  3031ff092d666d68  x5  3031ff092d666d68  x6  3031ff092d666d68  x7  7f7f7f7f7f7f7f7f
11-12 14:27:15.045  6909  6909 F DEBUG   :     x8  00000000000000f0  x9  0000007ebd86f278  x10 0000000000000001  x11 0000007ebd8b7a08
11-12 14:27:15.045  6909  6909 F DEBUG   :     x12 0000000000000000  x13 0000000000000000  x14 0000007bba963d80  x15 00000c55979a775c
11-12 14:27:15.045  6909  6909 F DEBUG   :     x16 0000007ebd91fd18  x17 0000007ebd8fb520  x18 0000007bb8d5e000  x19 000000000000171b
11-12 14:27:15.045  6909  6909 F DEBUG   :     x20 0000000000001afa  x21 00000000ffffffff  x22 0000000000001761  x23 000000000000171b
11-12 14:27:15.045  6909  6909 F DEBUG   :     x24 0000007bba966cb0  x25 0000007bba966cb0  x26 0000007bba966ff8  x27 00000000000fc000
11-12 14:27:15.045  6909  6909 F DEBUG   :     x28 00000000000fe000  x29 0000007bba9650c0
11-12 14:27:15.045  6909  6909 F DEBUG   :     lr  0000007ebd8a8704  sp  0000007bba965020  pc  0000007ebd8a8730  pst 0000000000001000
11-12 14:27:15.045  6909  6909 F DEBUG   : 18 total frames
11-12 14:27:15.045  6909  6909 F DEBUG   : backtrace:
11-12 14:27:15.045  6909  6909 F DEBUG   :       #00 pc 000000000005b730  /apex/com.android.runtime/lib64/bionic/libc.so (abort+168) (BuildId: 37f537c2ba9dcbb262a0a68f41a21da4)
11-12 14:27:15.045  6909  6909 F DEBUG   :       #01 pc 00000000029e5590  /data/app/~~H2tg_4OQQJKAw8po401eGw==/com.appname-lmQPfD9SCpEKSxsXXzpCmA==/base.apk!libappname.so (offset 0x308000) (absl::lts_20240116::log_internal::LogMessage::FailWithoutStackTrace()+108) (BuildId: 0543c9cb405df6dd7dcf35187e66232250e3b905)
11-12 14:27:15.045  6909  6909 F DEBUG   :       #02 pc 00000000029e5fcc  /data/app/~~H2tg_4OQQJKAw8po401eGw==/com.appname-lmQPfD9SCpEKSxsXXzpCmA==/base.apk!libappname.so (offset 0x308000) (absl::lts_20240116::log_internal::LogMessage::Die()+56) (BuildId: 0543c9cb405df6dd7dcf35187e66232250e3b905)
11-12 14:27:15.045  6909  6909 F DEBUG   :       #03 pc 00000000029e5920  /data/app/~~H2tg_4OQQJKAw8po401eGw==/com.appname-lmQPfD9SCpEKSxsXXzpCmA==/base.apk!libappname.so (offset 0x308000) (absl::lts_20240116::log_internal::LogMessage::SendToLog()+152) (BuildId: 0543c9cb405df6dd7dcf35187e66232250e3b905)
11-12 14:27:15.045  6909  6909 F DEBUG   :       #04 pc 00000000029e4cd4  /data/app/~~H2tg_4OQQJKAw8po401eGw==/com.appname-lmQPfD9SCpEKSxsXXzpCmA==/base.apk!libappname.so (offset 0x308000) (absl::lts_20240116::log_internal::LogMessage::Flush()+532) (BuildId: 0543c9cb405df6dd7dcf35187e66232250e3b905)
11-12 14:27:15.045  6909  6909 F DEBUG   :       #05 pc 00000000029e6168  /data/app/~~H2tg_4OQQJKAw8po401eGw==/com.appname-lmQPfD9SCpEKSxsXXzpCmA==/base.apk!libappname.so (offset 0x308000) (absl::lts_20240116::log_internal::LogMessageFatal::~LogMessageFatal()+20) (BuildId: 0543c9cb405df6dd7dcf35187e66232250e3b905)
11-12 14:27:15.046  6909  6909 F DEBUG   :       #06 pc 0000000001c639b0  /data/app/~~H2tg_4OQQJKAw8po401eGw==/com.appname-lmQPfD9SCpEKSxsXXzpCmA==/base.apk!libappname.so (offset 0x308000) (grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallNoOp<3>, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::ContinueFillOpsAfterInterception()+360) (BuildId: 0543c9cb405df6dd7dcf35187e66232250e3b905)
11-12 14:27:15.046  6909  6909 F DEBUG   :       #07 pc 0000000001c63790  /data/app/~~H2tg_4OQQJKAw8po401eGw==/com.appname-lmQPfD9SCpEKSxsXXzpCmA==/base.apk!libappname.so (offset 0x308000) (grpc::internal::CallOpSet<grpc::internal::CallOpSendInitialMetadata, grpc::internal::CallOpSendMessage, grpc::internal::CallNoOp<3>, grpc::internal::CallNoOp<4>, grpc::internal::CallNoOp<5>, grpc::internal::CallNoOp<6> >::FillOps(grpc::internal::Call*)+92) (BuildId: 0543c9cb405df6dd7dcf35187e66232250e3b905)
11-12 14:27:15.046  6909  6909 F DEBUG   :       #08 pc 0000000001c57d9c  /data/app/~~H2tg_4OQQJKAw8po401eGw==/com.appname-lmQPfD9SCpEKSxsXXzpCmA==/base.apk!libappname.so (offset 0x308000) (grpc::Server::PerformOpsOnCall(grpc::internal::CallOpSetInterface*, grpc::internal::Call*)+40) (BuildId: 0543c9cb405df6dd7dcf35187e66232250e3b905)
11-12 14:27:15.046  6909  6909 F DEBUG   :       #09 pc 00000000019d92c8  /data/app/~~H2tg_4OQQJKAw8po401eGw==/com.appname-lmQPfD9SCpEKSxsXXzpCmA==/base.apk!libappname.so (offset 0x308000) (grpc::internal::Call::PerformOps(grpc::internal::CallOpSetInterface*)+40) (BuildId: 0543c9cb405df6dd7dcf35187e66232250e3b905)
11-12 14:27:15.046  6909  6909 F DEBUG   :       #10 pc 00000000019e0fb0  /data/app/~~H2tg_4OQQJKAw8po401eGw==/com.appname-lmQPfD9SCpEKSxsXXzpCmA==/base.apk!libappname.so (offset 0x308000) (grpc::ServerWriter<C>::Write(C const&, grpc::WriteOptions)+316) (BuildId: 0543c9cb405df6dd7dcf35187e66232250e3b905)
<some more lines omitted)
11-12 14:27:15.046  6909  6909 F DEBUG   :       #16 pc 00000000000c37b4  /apex/com.android.runtime/lib64/bionic/libc.so (__pthread_start(void*)+208) (BuildId: 37f537c2ba9dcbb262a0a68f41a21da4)
11-12 14:27:15.046  6909  6909 F DEBUG   :       #17 pc 000000000005d084  /apex/com.android.runtime/lib64/bionic/libc.so (__start_thread+68) (BuildId: 37f537c2ba9dcbb262a0a68f41a21da4)
11-12 14:27:15.085  1832 29407 I sensors-hal: handle_sns_std_sensor_event:311, [SSC_LIGHT] P: 17(17),m:7,c:86,b:57,c0:1200,l:17,f(r):112,u(bl):100,j:14
```

### Anything else we should know about your project / environment?

**Comment by RubenGarcia** [comment]
My suspicion is that the
```
  return Status::OK;
```
from the function creating the thread is happening in parallel and causing the issue. I´m going to double-check if this is the issue.

**Comment by RubenGarcia** [comment]
I removed the thread creation and the crash disappears.
Since there is no way to know when A::B Status_OK is processed, that means that the writer is unusable in an external thread, and I will have to move to an asynchronous server.

I cannot share the code. I will try to create a simple example which can be shared.

**Comment by RubenGarcia** [comment]
I only called the rpc once, so there is no concurrent thread (except for main thread)

**Comment by yashykt** [comment]
I'm unable to confirm without looking at the rest of your code, but my suspicion is that you are invoking a `Write()` before the previous `Write()` completes. The API contract for `Write()` is that there can only be one outstanding `Write()` call at a moment.

**Comment by inferno-umar** [comment]
@yashykt I m too having same error but on client side grpc C++.

It happens when `grpc::ClientAsyncReaderWriter stream` on client writes data using `stream->Write()` when underlying channel is closed or reconnected!

Sometimes it crashes sometimes not!

**Comment by yashykt** [comment]
Hi, @inferno-umar, please share your code, for us to test in a new issue. I'm going to close this since this doesn't seem to be immediately actionable.
